merge updates

jskimerhorn opened the pull request:
merge updates to main BMIER repo

jskimerhorn committed:
new feature
the branches are now collapsable

jskimerhorn committed:
refactor
the plus and minus have been replaced by arrows

jskimerhorn committed:
changed the plus and minus sighns to an arrow icon

jskimerhorn committed:
fix bug
the message counting code was glitching when the first messages were added to a conversation.
added code to recognise this and return a value of 2

jskimerhorn committed:
fix bug
branch messages were being added in the wrong place. i added a plus one when i shouldnt have

jskimerhorn committed:
fix bug
the code for the dropdow was only returning the first character of each entry
because of this every conversation past 9 wasnt parsing branches correctly

jskimerhorn committed:
fix bug
the selected item id is index + 1 not - 1(oops)
a note on this has been added to the gui specification.
refactoring
the selected item has been removed from the arguments passed to the create branch function
it is no longer needed
due to being phased out in favor of the selected_item_id

jskimerhorn committed:
refactor
removed the cleaned branch messages function. it wasnt really adding anything
new feature
added config options for turing off the following features
summary
chat history
seperate conversations
branching
this will better allow adding modules that dont need these features

jskimerhorn committed:
refactor
removed the cleaned branch messages function. it wasnt really adding anything
new feature
added config options for turing off the following features
summary
chat history
seperate conversations
branching
this will better allow adding modules that dont need these features

jskimerhorn committed:
Merge branch 'collapsable-branches' of https://github.com/jskimerhorn/BMIER_project into collapsable-branches

jskimerhorn committed:
refactor
refactored the code for the if conversations are turned off to be more robust
added a bare bones kobold module

jskimerhorn committed:
fix bug
added code to abort adding a summary if the sumamry returned is blank
updated the lobold module to reture something if required
its still a placeholder but it actually works with the code
added a test ai module called test_module.py
its an exact copy of the kobold modue.
i caught several bugs so its staying around

jskimerhorn committed:
Merge pull request #3 from jskimerhorn/collapsable-branches

Collapsable branches

jskimerhorn committed:
refactor
moved the code for getting the model response to its own function to better allow modification
added the configs to the ai_database file.

jskimerhorn committed:
new feature
the first version of the kobold module is a success

jskimerhorn committed:
new feature
added a module for oobabooga

jskimerhorn committed:
Merge pull request #4 from jskimerhorn/kobold-module

Kobold module

jskimerhorn committed:
Merge pull request #5 from jskimerhorn/main

sync new changes

jskimerhorn committed:
fix bug
changed the port # for the browser gui to resolve a conflict with kobold
new feature
enabled the summary and chat history features on the kobold module

jskimerhorn committed:
fix bug
i removed the model from the summary because a particular model is not guaranteed to be installed

jskimerhorn committed:
fix bug
the summary config option wasnt actually doing anything
it now controls if a summary is made

jskimerhorn committed:
refactor
the code for parsing the response in the kobold module hjas bee optimized
specifically rhe part whewre the response is parsedf and formatted has been reduced to one line
the code for the summary config has been moved to the ai_datasbase module.
this way it doesnt need to be replicated for each module

jskimerhorn committed:
fix bug
renamed the config variable to avoid any potential conflicts caused by same named variables

jskimerhorn committed:
refactor
the code for chat history being enable has been moved to the ai_database module

jskimerhorn committed:
revert
the previous change to the chat history config didnt work so its been reverted

jskimerhorn committed:
fix bug
a warning message was added for if there is no model
changed some variable to be more uniqe

jskimerhorn committed:
fix bug
the program was crashing on submit when chat history was turned off
fix glitch
switching conversations when chat history was on would show the conversaion
new feature
turning off chat history now no longer logs the conversation in the database

jskimerhorn committed:
fix bug
modle wasnt parsing input when history was turned on

jskimerhorn committed:
fix bug
if the model is not loaded the message from the user will be ignored
the modle not being present wasnt being properl;y detected in oobabooga
added some code to simply detect if the response was blank

jskimerhorn committed:
refactor
modifications made to the submit function to clean it up

jskimerhorn committed:
fix bug
the config file has been deleted. it shouldnt be tracked

jskimerhorn committed:
config.json added to the git ignore properly
default config file added

jskimerhorn committed:
fix bug/ refactor
fixed multiple bugs with the  kobold/ oobabooga modules
the model is now called bot instead of assistant when being sent to the model.
the model is still assistant in the database
modified the chat memory for the modules to strip to work better with the backends

jskimerhorn committed:
Merge pull request #6 from jskimerhorn/bug-fixes

Bug fixes

jskimerhorn committed:
moved the ai and database modules to the modules subfolder.
removed the direct import for the database module.
changd the wording for an error message to be more clear

jskimerhorn committed:
modified an error message to be clearer

smytjf11 commented on ai_database.py:
```diff
@@ -12,54 +12,68 @@ def load_config():
     return config
 
 config = load_config()
+chat_history_enabled = config['chat_history']
+summary_enabled = config['summary']
+branching_enabled = config['branching']
+conversations_enabled = config['conversations']
```
```suggestion
chat_history_enabled, summary_enabled , branching_enabled, conversations_enabled = \
    config[['chat_history', 'summary', 'branching', 'conversations']]
```

jskimerhorn replied:
```suggestion
chat_history_enabled, summary_enabled, branching_enabled, conversations_enabled = \
    config['chat_history'], config['summary'], config['branching'], config['conversations']


```
good catch. i had to do some bug fixing but i got it to work.

smytjf11 commented on ai_database.py:
```diff
@@ -12,54 +12,68 @@ def load_config():
-module_name = f"{model.replace('.', '_')}_module"
-ai_module = importlib.import_module(f"{module_name}")
-
+model_module_name = f"{model.replace('.', '_')}_module"
+# the location of the module is now in a subfolder \modules\ai_model\
+ai_module = importlib.import_module(f"modules.ai_model.{model_module_name}")
 
 gui = config['gui']
-module_name = f"{gui.replace('.', '_')}_gui"
-gui_module = importlib.import_module(f"{module_name}")
+gui_module_name = f"{gui.replace('.', '_')}_gui"
+gui_module = importlib.import_module(f"{gui_module_name}")
 
 
 database = config['database']
-module_name = f"{database.replace('.', '_')}_database"
-database_module = importlib.import_module(f"{module_name}")
+database_module_name = f"{database.replace('.', '_')}_database"
+database_module = importlib.import_module(f"modules.database.{database_module_name}")
```
Could we take this directly from the config?

jskimerhorn replied:
yes we can. i added a pull request for the changes due to it modifying multiple files

smytjf11 commented on ai_database.py:
```diff
@@ -73,18 +87,24 @@ def fetch_chat_history(self, conversation_id):
 def send_to_api(self, input_text, conversation_id, selected_item, selected_branch_conversation_id=None):
     # Fetch the chat history lines from the parent conversation
-    chat_history_lines = fetch_chat_history(self, conversation_id)
+    if chat_history_enabled == True:
```
```suggestion
    if chat_history_enabled:
```

jskimerhorn replied:
i didnt know you could do thast. it looks a lot cleaner. i will keep this in mind

smytjf11 commented on ai_database.py:
```diff
@@ -73,18 +87,24 @@ def fetch_chat_history(self, conversation_id):
 
     # Fetch the messages from the selected branch (if any) and convert them to a list of dictionaries
-    if selected_branch_conversation_id:
+    if branching_enabled == True and selected_branch_conversation_id:
```
```suggestion
    if branching_enabled and selected_branch_conversation_id:
```

smytjf11 commented on ai_database.py:
```diff
@@ -73,18 +87,24 @@ def fetch_chat_history(self, conversation_id):
+    combined_messages = chat_history_lines + branch_messages
+    chat_memory = ai_module.construct_chat_memory(self, input_text, combined_messages)
+    response = ai_module.get_response(self, chat_memory)
+    # check if the response is empty
```
```suggestion
    # TODO: check if the response is empty
```

jskimerhorn replied:
i dont know why this comment is here. i say we get rid of it
```suggestion

```

smytjf11 commented on default_config.json:
```diff
@@ -0,0 +1,12 @@
+{
```
We should comment on what these inputs do and their default/allowable values

jskimerhorn replied:
apparent .json doesnt use comments.

smytjf11 commented on desktop_gui.py:
```diff
@@ -26,19 +24,20 @@ def load_config():
-print(" summary update interval is ", config['summary_update_interval'])
+chat_history_enabled = config['chat_history']
+branching_enabled = config['branching']
+conversations_enabled = config['conversations']
```
For cases like this where we are using the same blocks of code/making the same changes, we can define an initialization function we can call at the top/main section of each script to streamline things and for quality of life

smytjf11 commented on desktop_gui.py:
```diff
@@ -85,28 +84,52 @@ def initUI(self):
+            # connect the signal of the dropdown menu to the switch conversation function
+            self.conversation.currentTextChanged.connect(self.switch_conversation)
+        else:
+            # if conversations are not enabled, but history is enabled, we need to set the conversation id to the first conversation id
```
I'm not sure how we want to handle this case, does it make sense to have history enabled but conversations not? What history is it supposed to remember?

jskimerhorn replied:
conversation is for whether of not there is a dropdown. history is for whether or not the front end remembers past conversational turns. in the case of conversation being off but history remaining one i currently have it set to use id 0. it was the easiest to impliment at the time

smytjf11 commented on desktop_gui.py:
```diff
@@ -151,11 +175,23 @@ def switch_conversation(self):
         conversation_id = selected_conversation_id
         
-
+    def warn(self, message):
+            msg = QtWidgets.QMessageBox()
+            msg.setIcon(QtWidgets.QMessageBox.Warning)
+            if message == "no model loaded":
+                msg.setText("The request failed. this is likely due to the model not being loaded. this module requires the "
+            + model + " api to be running. please check that the kobold api is running and try again.")
+            if message == "blank":
+                msg.setText("The request failed. this is likely due an error causing the response from the model to be blank.")
+            msg.setWindowTitle("Warning")
+            msg.setStandardButtons(QtWidgets.QMessageBox.Ok)
+            msg.exec_()
+            return
```
Here would be a good spot to add logging, pretty much exactly as it's being written to the app

smytjf11 commented on test_module.py:
```diff
@@ -0,0 +1,58 @@
+# this module is a test module to allow for testing of the program without the need for the gpt3.5 model
```
Recommend we change this to Unit Testing and start implementing those!

jskimerhorn replied:
this module was used to test the kobold module befor i implimented it. this entire file can be safely deleted

smytjf11 reviewed (commented):
We should consider adding docstrings to the functions, Copilot/ChatGPT/Bing should be able to help with that, or some other model

jskimerhorn committed:
Merge pull request #7 from smytjf11/main

Merge pull request #6 from smytjf11/mobile-gui

jskimerhorn committed:
implemented suggested changes

jskimerhorn committed:
first draft of module

jskimerhorn committed:
deleted test module.
it was no longer needed

jskimerhorn committed:
refactoring
uncommented the try accept block in the get_response function
changewd output to assistant message so that we arent reusing the response variable
deleted unneeded commented out code in the construct_chat_memory function

jskimerhorn committed:
fix bug
added back the _get_history function
it has been renamed convert_history_format
this better reflects its actual purpose
added max_tokens and stop parameters to the model api call in the oobabooga module
this prevents the model from terminating at 16 tokens

jskimerhorn committed:
refactor
moved the max tokens and stop sequence options for the ooba module to the congig file.

jskimerhorn committed:
fix bug
the new chat button is now disabled if conversations are disabled in the config file
the summar feature now works with the new obbabooga module
the default config now uses the ooba booga module

jskimerhorn committed:
fix bug
added code to prevent a conversation id from being used twice if the new conversation button was pressed again befor anything was added to the conversation

jskimerhorn committed:
fix bug
the code wasnt adding the assistant and user lables to the text in the ui  untill the conversation was reloaded

jskimerhorn committed:
fix bug
the assistant and user lables werent showing up on newly added messages
added code to add them to each message as it is recieved from the model
